### PR TITLE
 CTCTRADERS-1357: Guarantee not valid message

### PR DESCRIPTION
**Pull request by ashwinirudrappa**

### CODE REVIEW

File: app/connectors/DepartureMovementConnector.scala [hunk 1]
```diff
@@ -33,4 +37,23 @@ class DepartureMovementConnector @Inject()(val appConfig: FrontendAppConfig, htt
...
+    }
+  }
+
+  def getGuaranteeNotValidMessage(location: String)(implicit hc: HeaderCarrier): Future[Option[GuaranteeNotValidMessage]] = {
+    val serviceUrl = s"${appConfig.departureBaseUrl}$location"
+    http.GET[HttpResponse](serviceUrl) map {
+      case responseMessage if is2xx(responseMessage.status) =>
+        val message: NodeSeq = responseMessage.json.as[ResponseMessage].message
+        XmlReader.of[GuaranteeNotValidMessage].read(message).toOption
+      case _ => None
```
Comment (chris-ianson) [inline app/connectors/DepartureMovementConnector.scala:56]: Could we put a log when we fail to get a message back.

File: app/connectors/DepartureMovementConnector.scala [hunk 2]
```diff
@@ -33,4 +37,23 @@ class DepartureMovementConnector @Inject()(val appConfig: FrontendAppConfig, htt
 
     http.POSTString[HttpResponse](serviceUrl, departureMovement.toXml.toString, headers)
   }
+
+  def getSummary(departureId: DepartureId)(implicit hc: HeaderCarrier): Future[Option[MessagesSummary]] = {
+
+    val serviceUrl: String = s"${appConfig.departureHost}/movements/departures/${departureId.value}/messages/summary"
+    http.GET[HttpResponse](serviceUrl) map {
+      case responseMessage if is2xx(responseMessage.status) => Some(responseMessage.json.as[MessagesSummary])
+      case _                                                => None
```
Comment (chris-ianson) [inline app/connectors/DepartureMovementConnector.scala:46]: Could we put a log when we fail to get a message back.

File: app/services/GuaranteeNotValidMessageService.scala [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+class GuaranteeNotValidMessageService @Inject()(connectors: DepartureMovementConnector) {
+
+  def guaranteeNotValidMessage(departureId: DepartureId)(implicit hc: HeaderCarrier, ec: ExecutionContext): Future[Option[GuaranteeNotValidMessage]] =
+    connectors.getSummary(departureId) flatMap {
+      case Some(summary) =>
+        summary.messagesLocation.guaranteeNotValid match {
+          case Some(location) => {
+            connectors.getGuaranteeNotValidMessage(location)
+          }
+          case _ => Future.successful(None)
```
Comment (chris-ianson) [inline app/services/GuaranteeNotValidMessageService.scala:35]: Some logs on the None would be useful

File: test/models/InvalidGuaranteeReasonCodeSpec.scala [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+              <GuaRefNumGRNREF21>{reasonCode.guaranteeRefNumber}</GuaRefNumGRNREF21>
+              <INVGUARNS>
+                <InvGuaReaCodRNS11>{reasonCode.code.value}</InvGuaReaCodRNS11>
+                ++ {reasonOptXml}
+              </INVGUARNS>
+            </GUAREF2>
+
+          XmlReader.of[InvalidGuaranteeReasonCode].read(xml).toOption.value mustBe reasonCode
+      }
+    }
```
Comment (chris-ianson) [inline test/models/InvalidGuaranteeReasonCodeSpec.scala:46]: Could we have another test in here that checks an unsupported code is still parsed to DefaultCode

**Review by chris-ianson — CHANGES_REQUESTED** [review]
Looks. Good, couple of comments. Do you think we should log out when an unsupported code is returned? Is there any value in that? (would we need to do anything if guarantee rejections are coming back with codes we don't know about)

File: test/models/InvalidGuaranteeReasonCodeSpec.scala (hunk 4 above)
Comment (ashwinirudrappa) [reply test/models/InvalidGuaranteeReasonCodeSpec.scala:46]: @chris-ianson I think this test covers the mentioned case, you want me to put the arbitrary String

Comment (chris-ianson) [reply test/models/InvalidGuaranteeReasonCodeSpec.scala:46]: I'm thinking of a test which covers a code outside of G01-G13, so we know the DefaultCode is tested. So keep what you have but have an additional test which sets _InvGuaReaCodRNS11_ to a code not supported

**Review by chris-ianson — APPROVED** [review]

**Pull request merged**